### ISSUE TITLE
http cronjob in a multi-account setup able to execute recurring actions with wrong command line token

### ISSUE THREAD

**Issue by BoudewijnSwen** [firefly-iii/firefly-iii#4484]
**Bug description**
I am running Firefly III version 5.4.6 in Kubernetes. I have two accounts setup, one for my personal finance and one for my partner. I have been testing with setting up a k8s cronjob for running the FF cron to execute the recurring transactions etc. 

**Steps to reproduce**
The yaml for the cronjob is as follows:
```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: firefly-cron-job
  namespace: firefly-iii
spec:
  schedule: "0 3 * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - image: radial/busyboxplus:curl
            name: firefly-cron
            args:
            - curl
            - http://10.13.37.98/cron/run/<command line token account 2>
```

Not sure if this is relevant, but one is admin, the other is not:
_**'Account 1'** = Is admin
**'Account 2'** = Is not admin_
I added the Command Line token from **account 2** (not admin) to the url for the cronjob.
I also created a recurring transaction, first under **account 2**, then another under account 1. Both the same, both running everyday.

The morning after I checked if the recurring transfer was created. I found that only the recurring transfer for **account 1** was created! 
- I ran curl http://10.13.37.98/cron/run/<command line token **account 2**> manually again, now the recurring transfer for **account 2** was created as well.
- I slightly editted the test recurring transfer under **account 2**.

The morning after only the recurring transfer from **account 2** was created.
- I ran curl http://10.13.37.98/cron/run/<command line token **account 1**>, output:
```
The recurring transaction cron job fired successfully.
The auto budget cron job fired successfully.
The telemetry cron job did not fire.
```
- Now the recurring transfer for **account 1** was created as well.

**Extra**
- On both days ran the curl command manually a 2nd time:
```
The recurring transaction cron job did not fire.
The auto budget cron job did not fire.
The telemetry cron job did not fire.
```
But this output is expected I guess, both accounts had their cronjobs ran. 

**Conclusion**
It seems only for one of the accounts the cronjobs get executed per trigger over http. No mater which command line token you use, only the recurring transactions that were editted last, get executed. 

**Solution/work around**
I will try and just schedule a 2nd cronjob to run the curl command with the command line token for the other account, so if and when this gets fixed, it wont break my setup :)

**Final thoughts**
This is nothing major I guess, but I though to report this behavior anyway, cause in my mind this does not feels like expected behavior that I am able to run cronjobs for account 2 from a account 1 token, and also that it only triggers the last one edited.
However, I have to say: I love firefly, I have been having a great learning experience setting it up in Kubernetes. Hopefully I might be able to contribute one day! Keep up the good work!

**Comment by JC5** [comment]
Thanks for the extensive report! I'll see what I can do re: the firing of the cron job. I have to do a few deep dives on the code on how it works exactly.

The cli token for the cron job is only there to make sure that it's not fired by random people / spammed. The idea is that the cron job will fire for all users regardless of what token is used. I was under the impression that the code could handle this, but clearly this is not the case.

**Comment by BWohlbrecht** [comment]
In my opinion the command line tokens should only be able to run the tasks in their specific account. I would expect that bevaviour.

Conceptually wouldn't it be better to introduce a new "super-token", which has the ability to trigger actions on all accounts?
Maybe give admins an option to generate such a token?

**Comment by JC5** [comment]
I can't replicate this. Whatever I do, everything is created as it should be.

If you run into this again, I'd appreciate some log files to see what happens exactly.

**Comment by BoudewijnSwen** [comment]
I will try to do some testing this week and get some logs.
Is there a easy to test with triggering cron's multiple times a day? Cause I had to wait a day every time I made changes to see what happens.

**Comment by BoudewijnSwen** [comment]
I doubt that will work with a http request though ;) But i'll see what I can do. I have redeployed my own environment ones of twice since creating this issue. So I hope this was not a fluke...

**Comment by BoudewijnSwen** [comment]
Debug information generated at 2021-04-01 09:54:28 UTC for Firefly III version **5.4.6**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.4.6 |
| Firefly III API | 1.4.0 |
| PHP | 7.4.14 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | 975be5a4-c011-4b9f-b71c-da87db87a4d1 |
| Using docker? | true |
| Telemetry | false |
| Layout |  |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | info, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| DB version | 15 (exp. 15) |
| Login provider | eloquent  |
| Trusted proxies (.env) | ** |

| User info | Value |
| --- | --- |
| Session start |  |
| Session end |  |
| Session first |  |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36 |


Sorry for the late response, I have not been able to do a lot of testing so far.
I have about 20~ recurring transactions on both accounts and they should have fired today. I run the container twice, ones with command line token voor account 1 and ones with the token for account 2. Only 2 recurring transactions were created, both on account 1. But this seems to be intended, that you fire all transactions on all accounts, not mather which account the command line token is from. So when I dove into why only 2 transactions were created, I found that I ran into the same bug as #4070 

Part of the logs of the container:
```
<div class="row">
    <div class="col-lg-10 col-lg-offset-1 col-md-12 col-sm-12 col-xs-12">
        <h3 class="text-info">Whoops! An error occurred.</h3>
    </div>
</div>
<div class="row">
    <div class="col-lg-10 col-lg-offset-1 col-md-12 col-sm-12 col-xs-12">
        <p>
            Unfortunately, this error was not recoverable :(. Firefly III broke. The error is:
        </p>
        <p class="text-danger">
            Typed property FireflyIII\Repositories\Tag\TagRepository::$user must not be accessed before initialization

        </p>
    </div>
</div>
```
If I manually curl http://10.13.37.98/cron/run/<token> it adds one more transaction and then errors out again. I'll just have to schedule this container 40 times for now. Not pretty, but he, if it works ;) 

I am really busy at work right now, but I will try and run some tests with the actual topic of this issue this weekend. 
Namely, create a (non tagged!) recurring transaction under both accounts and try to trigger them via a single curl command. See if they both get created or not. And of course produce log output!

**Comment by JC5** [comment]
Ah, that could explain the original issue. Thanks for getting back to me! I can do some research and include a fix in 5.5.2.
